### ISSUE TITLE
Change the directional bias of prompts

### ISSUE THREAD

**Issue by sjdrodge** [hanabi/hanabi.github.io#677]
Currently we play w/ a bias towards prompting newer cards (leftmost). I'm proposing that we change to an older card bias (rightmost).

The reason is that older cards by definition always have at least as much negative information on them as newer cards do.

If we make this change we get the following prompt properties for free:
- Known rainbow cards prompt before possibly rainbow cards when you are looking for a rainbow card. (this is already the convention, as an explicit exception to the default prompt rule)
- Known non-rainbow cards prompt before possibly rainbow cards when you are looking for a non-rainbow card. (not currently the convention, considered to be too easy to forget)
- Known non-omni cards prompt before possibly omni cards when you're looking for a non-omni card. (not currently the convention, but imo a highly desirable addition even if this proposal is not accepted wholesale)
- Known omni cards prompt before possibly omni cards when you're looking for an omni card. (somewhat ambiguous whether this is currently the convention)

The only downside that I can imagine is if there's some unaccounted for reason why leftmost cards are more likely to be in prompt position. If anyone knows a reason that this would be true (beyond players setting up future prompts due to their knowledge of the current convention), please let me know.

**Comment by Lel0uch-H** [comment]
Agree with the overall idea, although it is so radically opposed to the status quo that I have little hope it will be accepted. :D

Additionally, the issue of cards having more negative info being prompted in rainbow/omni variants is something I wholeheartedly agree with and I have even seen some players play like this from intuition. Currently though, the 'negative prompt' convention only exists in the 'Dual-Color (6 suits)' variant-specific conventions, I would really like if it were extended to all variants.

**Comment by Pond26535** [comment]
Ironically this used to be the standard convention, however things have changed, Its actually the only convention I've play with for an significant amount of time.  The disadvantage that hasn't been mentioned is it makes many chop cluing techniques redundant as you are already cluing the chop often.

**Comment by aliblong** [comment]
One small additional benefit is that tempo clues remaining left-to-right would increase convention coverage in situations where both a prompt and tempo clue exist. Imagine Bob holds b234 clued as blue and Cathy holds b5. Alice can choose whether to tempo clue Bob's blue cards or clue Cathy's b5 depending on whether the b2 is left or right.

**Comment by pianoblook** [comment]
> The reason is that older cards by definition always have at least as much negative information on them as newer cards do.

That's not always true though. At least a good chunk of the time I suspect whichever 3 was *focused* will tend to have more negative info.

For example: 4p No Variant, turn 1:
- Alice clues Bob three 1s: r,y,g. Bob plays g1
- Cathy clues 3 to Alice, touching slot 1 and 3
- Donald immediately blind-plays a b1, as a 3-bluff
- Alice's leftmost 3 now has negative r+y, while the rightmost one can be any

**Comment by pianoblook** [comment]
> Alice can choose whether to tempo clue Bob's blue cards or clue Cathy's b5 depending on whether the b2 is left or right. This effect ends up being more powerful for suits where it's tough to get precise information.

This is pretty compelling though

**Comment by sjdrodge** [comment]
> > The reason is that older cards by definition always have at least as much negative information on them as newer cards do.
> 
> That's not always true though. At least a good chunk of the time I suspect whichever ~3~ cards were previously _focused_ will tend to have more negative info.
> 
> For example: 4p No Variant, turn 1:
> 
> * Alice clues Bob three 1s: r,y,g. Bob plays g1
> * Cathy clues 3 to Alice, touching slot 1 and 3
> * Donald immediately blind-plays a b1, as a 3-bluff
> * Alice's leftmost 3 now has negative r+y, while the rightmost one can be any

Indeed, my statement is only absolutely true for non-contextual negative information.

**Comment by florian-5f3759df** [comment]
It sounds really reasonable to me. It will take some getting used to, but I'm totally willing to test it whenever playing with any of you and see how it feels.

Small side note: Brown tempo clues probably will have to be from the left if this is passed. (And then I guess if nobody tempos your two brown cards, and n4 n5 are missing, you might play from the right by default then, instead of the left.)

**Comment by sjdrodge** [comment]
Agreed about the brown changes. What is nice is that brown is no longer an exception then. It works the same as every other suit.

**Comment by fpvandoorn** [comment]
My main concern is that wanting to save a card negatively correlates with wanting to prompt a card. If I get a clue on chop, and I happen to know for a fact that it was 100% a save clue, that means that a prompt immediately afterwards, it is not on the saved card. 

Of course, this exception is baked in: if you know that the prompt is not on the rightmost card, then it is on the leftmost card. But suppose the clue was 4, and soon after one critical 4 becomes playable. I think on average it is still more likely that the leftmost 4 is playable than the rightmost (saved) card.

**Comment by Zamiell** [comment]
agree with mr fpvandoorn

**Comment by sjdrodge** [comment]
> My main concern is that wanting to save a card negatively correlates with wanting to prompt a card. If I get a clue on chop, and I happen to know for a fact that it was 100% a save clue, that means that a prompt immediately afterwards, it is not on the saved card.
> 
> Of course, this exception is baked in: if you know that the prompt is not on the rightmost card, then it is on the leftmost card. But suppose the clue was 4, and soon after one critical 4 becomes playable. I think on average it is still more likely that the leftmost 4 is playable than the rightmost (saved) card.

I think this is the major "unaccounted for reason" that I asked for. More testing is required to find out how large this effect actually is. My personal largest concern currently is w/ 2 saves in variants with high clue fidelity (e.g. No Variant), but as of now I'm not sure how much of this downside is "real" and how much is just that our current clue choices are geared towards leftmost prompt bias.

So far, my estimation is that this prompt rule shines in Omni, might be a slight improvement in Rainbow, and is almost irrelevant in most other variants. Haven't tested brown yet, but I suspect it may be an improvement there as well.

**Comment by Zamiell** [comment]
> all current prompt exceptions

can you enumerate those

**Comment by sjdrodge** [comment]
> > all current prompt exceptions
> 
> can you enumerate those

Keeps the same behavior here:
- https://hanabi.github.io/docs/level-5/#prompts-in-multi-color-variants
- https://hanabi.github.io/docs/extras/miscellaneous/#double-prompts-in-multi-color-variants
- https://hanabi.github.io/docs/variant-specific/dual-color-6-suits/#the-negative-prompt

Changes the behavior here:
- https://hanabi.github.io/docs/variant-specific/pink/#prompts-with-pink-promise

**Comment by Zamiell** [comment]
I like reducing complexity / deleting conventions.
With that said, deleting variant-specific conventions is not a very big win for us.
Specifically, the goal of our conventions is to get the max win-rate in No Variant (and Rainbow-6S / Blank-6S).
Thus, making modifications for the purposes of simplifying exotic variants like Omni is a **non-factor**, or at the very least a minor factor that should not be the main focus.
In other words, there has to be *some other benefit* that makes the proposed reconfiguration better for NV on average.
With what Florrat said in mind, is NV better or worse on average? It sounds like it would be worse.

**Comment by sjdrodge** [comment]
For NV it's slightly worse, for Rainbow it's slightly better. Making the convention system generalize better across variants should be considered a significant win.

**Comment by timotree3** [comment]
Will we have to change the Leftmost Playable Principle?

**Comment by Zamiell** [comment]
yes, such a reconfiguration would change the first principles

**Comment by Zamiell** [comment]
given the most recent discussion, does anyone else feel strongly either way

**Comment by Zamiell** [comment]
for what its worth, i don't care very much about making the convention system generalize well across exotic variants, and i don't think the group should care either.
i could invent more exotic variants tomorrow, or choose to delete exotic variants tomorrow.
both are possible, but my arbitrary decisions relating to the website should not affect the way that the group plays No Variant.

furthermore, considering the friction in changing the *First Principles*, i would prefer that we get some major benefit other than "NV it's slightly worse, for Rainbow it's slightly better".

**Comment by sjdrodge** [comment]
I'm disturbed by the suggestion that variants like pink and omni are candidates for deletion. I really hope that suggestion is just an argument technique and not an accurate representation of what might actually come to pass.

To be clear, this change removes at least 5 special exceptions that we currently do play, and it also gives us several that we know are good but don't currently play due to fear of complexity. I think it's a major flaw in the design goals of the conventions if a change that creates a substantial simplification across all known variants is rejected on the basis that it isn't gaining anything for No Variant. Also, I have a feeling that if I offered a set of convention changes that increased NV winrate by ~3% but made the conventions no longer generalize well to other variants, that this design maxim might suddenly disappear (I do actually have a few sets of changes that would do exactly that, so this is not purely theoretical).

That said, the actual increase in convention effectiveness is quite small from this change. Prompts with multiple "candidate cards" come up much less frequently than I had realized before trying to test this. Therefore, I think it's perfectly reasonable to reject this on the basis that changes this intrusive should gain a lot of edge, and this one does not.

**Comment by Zamiell** [comment]
> I'm disturbed by the suggestion that variants like pink and omni are candidates for deletion

pink and white are non-cannon and will always be candidates for deletion, but omni and null would be deleted first before that

> I have a feeling that if I offered a set of convention changes that increased NV winrate by ~3% but made the conventions no longer generalize well to other variants, that this design maxim might suddenly disappear

that entirely depends on how much it doesn't generalize anymore!
as i said above, its *nice* if the conventions generalize to all different kinds of variants, but that isn't an explicit design goal

to repeat myself a little bit, my high-level point is that if:
- i add a Zulu variant
- and 25% of Hyphenated members hate it and don't play it
- and 75% of Hyphenated members play it
- and someone proposes convention X that will make zulu better, while simultaneously shifting around no variant conventions
- then the 25% of the people who don't play zulu and only play no variant are going to be pissed off
- furthermore, if i decide to delete zulu in the future, then convention X will essentially be for nothing, and we will create more friction by changing everything back to the way it was

i feel that this hypothetical is grounded in reality, since players like ahming continue to play with the group and hardly play any exotic variants because they prefer NV/rainbow

**Comment by sjdrodge** [comment]
> pink and white are non-cannon and will always be candidates for deletion, but omni and null would be deleted first before that

Given that most of the H-group plays quite a lot of these variants, the fact that this is under consideration is disturbing to me.

The hypothetical you lay out bears little resemblance to the real situation, but I certainly agree that if we were in the hypothetical situation, your argument would apply.

**Comment by iamwhoiamhahaha** [comment]
Can we reopen this issue? @Zamiell 

https://discord.com/channels/140016142600241152/1224898857843032105/1226025734712201266

> pianoblook (they/he) — 04/03/2024
> fwiw KRU has adopted rightmost prompt, and it's been heavenly. A couple benefits:
> Bob will always have greater than or equal information on the older card (particularly nice in pink/rainbow suits)
> Synergizes well with (direct) tempo clues; if you keep those focused leftmost, it feels nice to have both options available sometimes.
> The downside has just been adjusting to it, lol, but you get used to it pretty quick.
> 
> We've also adopted rightmost kt discards as well, which is fairly inconsequential but can have some nice positional-play benefits nearing lategame compared to discarding newest (i.e. since it's a lot less common to want to positional discard for Bob's rightmost than his leftmost)

> IAMJEFF — Today
I personally think that prompting from right is more logical because they have more negative information. The reason why prompts take precedence over finesses is that the probability of the clued card which matches to the clue is higher than the probability of the card which is on finesse position. Let's say Bob has two clued 3s. One of them is purely unknown and the other has all negative information except for red and green. Alice play clues a red 4. I would say it's more logical for Bob to play the 3 that can only be red or green rather than the 3 which is purely unknown because of probability. Another reason why prompting from right is better is that when someone gives a tempo clue, it's usually better to touch the leftmost clued card rather than the rightmost. It's because the likelihood the player might be able to know the identity of their rightmost clued card is greater than the leftmost because of negative information. And sometimes, filling the leftmost card would immediately reveal the identity of the rightmost card. So what's the point to tempo a card which is on the rightmost? This concept coheres with Tempo Clue Chop Moves. Because directly tempoing a clued card on the rightmost card is so bad, it should at least chop move one card. For discarding trash from right, I don't think it conflicts with gentleman's discards. Imagine we're in a world where h-group players discard trash from right. Everyone should understand why the player discarding trash from left in a gentleman's discard situation and it doesn't chop move. Similar to why doing a gentleman's discard while holding a playable card doesn't shout/scream to the next player. It's because everyone knows what's going on and that's it.

> IAMJEFF — Today
Here I provide an example explaining why Prompting from Right is better. Let's say someone has four 4s with positive rank clues on them, and they have different negative color information as following:
rygbp[4], rygb[4], rgb[4], rg[4]
In Prompting from Left system, only rygbp[4] is promptable; while in Prompting from Right system, all of the cards are promptable. In Prompting from Right system, rgb[4] can be prompted by a b5 without being blocked by rg[4]. While rygbp[4] is always blocking in Prompting from Left system.

I've been pondering this issue for some time now, and I strongly advocate for changing the way how we prompt a card.

> [fpvandoorn](https://github.com/fpvandoorn) commented [on Jul 3, 2021](https://github.com/hanabi/hanabi.github.io/issues/677#issuecomment-873268601)
My main concern is that wanting to save a card negatively correlates with wanting to prompt a card. If I get a clue on chop, and I happen to know for a fact that it was 100% a save clue, that means that a prompt immediately afterwards, it is not on the saved card.

Good point! But I think it's easy to fix this problem. How about:

### Prompt Order
* The leftmost card with the most global negative information.
* For example:
  * Bob has five 4s in his hand: `rygbp[4], rgb[4], rgb[4], rg[4], rg[4]`
  * The *Prompt Order* should be: slot 4, slot 5, slot 2, slot 3, and then slot 1

**Comment by Zamiell** [comment]
jeff asked me to reopen on april 6th, but there hasn't been any discussion whatsoever since then, so i'll close this once again
